app/models/admin_dashboard_data.rb

```diff
@@ -58,49 +74,148 @@ def problems
     problems
   end
 
-  def self.problems_started_key
-    'dash-problems-started-at'
+  def self.add_problem_check(*syms, &blk)
+    @problem_syms.push(*syms) if syms
+    @problem_blocks << blk if blk
   end
 
-  def self.set_problems_started
-    existing_time = Discourse.redis.get(problems_started_key)
-    Discourse.redis.setex(problems_started_key, 14.days.to_i, existing_time || Time.zone.now.to_s)
+  def self.add_scheduled_problem_check(check_identifier, &blk)
+    @problem_scheduled_check_blocks[check_identifier] = blk
   end
 
-  def self.clear_problems_started
-    Discourse.redis.del problems_started_key
+  def self.add_found_scheduled_check_problem(problem)
+    problems = load_found_scheduled_check_problems
+    if problem.identifier.present?
+      return if problems.find { |p| p.identifier == problem.identifier }
+    end
+    problems << problem
+    set_found_scheduled_check_problems(problems)
   end
 
-  def self.problems_started_at
-    s = Discourse.redis.get(problems_started_key)
-    s ? Time.zone.parse(s) : nil
+  def self.set_found_scheduled_check_problems(problems)
+    Discourse.redis.setex(SCHEDULED_PROBLEM_STORAGE_KEY, 300, JSON.dump(problems.map(&:to_h)))
+  end
+
+  def self.clear_found_scheduled_check_problems
+    Discourse.redis.del(SCHEDULED_PROBLEM_STORAGE_KEY)
+  end
+
+  def self.clear_found_problem(identifier)
+    problems = load_found_scheduled_check_problems
+    problems.reject! { |p| p.identifier == identifier }
+    set_found_scheduled_check_problems(problems)
+  end
+
+  def self.load_found_scheduled_check_problems
+    found_problems_json = Discourse.redis.get(SCHEDULED_PROBLEM_STORAGE_KEY)
+    return [] if found_problems_json.blank?
+    begin
+      JSON.parse(found_problems_json).map do |problem|
+        Problem.from_h(problem)
+      end
+    rescue JSON::ParserError => err
+      Discourse.warn_exception(err, message: "Error parsing found problem JSON in admin dashboard: #{found_problems_json}")
+      []
+    end
   end
 
-  # used for testing
+  def self.register_default_scheduled_problem_checks
+    # TODO (martin) Add group SMTP check here
+  end
+
+  def self.execute_scheduled_checks
+    found_problems = []
+    problem_scheduled_check_blocks.each do |check_identifier, blk|
+      problems = nil
+
+      begin
+        problems = instance_exec(&blk)
+      rescue StandardError => err
+        Discourse.warn_exception(err, message: "A scheduled admin dashboard problem check (#{check_identifier}) errored.")
+        # we don't want to hold up other checks because this one errored
+        next
+      end
+
+      found_problems += Array.wrap(problems)
+    end
+
+    found_problems.compact.each do |problem|
+      next if !problem.is_a?(Problem)
+      add_found_scheduled_check_problem(problem)
+    end
+  end
+
+  class << self
+    attr_reader :problem_syms,
+                :problem_blocks,
+                :problem_messages,
+                :problem_scheduled_check_blocks
+  end
```

Comment:
I think [mattr_reader](https://api.rubyonrails.org/classes/Module.html#method-i-mattr_reader) will work here.

Reply:
What is the benefit of this? Can `mattr_reader` be used on the `AdminDashboardData` class itself?

Reply:
O this is just some Rails magic that allows us to define class attributes. Without having to do the `class << self` dance. No real benefit aside from less code to write but I do see it being used alot in the Rails code base.

> Defines a class attribute and creates a class and instance reader methods.

`cattr_reader` is an alias to `mattr_reader`.

Reply:
Ah okay I see, thank you!